Global Load Balancer Pod is crashing

Hi @raffaelespazzoli ,

We are trying to use the Global Load Balancer operator in our environment. When we tried to create the Global DNS Record, the pod is getting crashing. Please help us to figure out the issue. We have troubleshoot the code and we found that is is failing in `getIPs`. Please check the below log.

```
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x30 pc=0x1800cd4]

goroutine 263 [running]:
k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
	/go/pkg/mod/k8s.io/apimachinery@v0.20.0/pkg/util/runtime/runtime.go:55 +0x10c
panic(0x19d1e80, 0x2a188c0)
	/usr/local/go/src/runtime/panic.go:969 +0x1b9
github.com/redhat-cop/global-load-balancer-operator/controllers/globaldnsrecord.(*EndpointStatus).getIPs(0xc006ee9b00, 0xc006eecfd0, 0xc006ee7f98, 0xc0049bb320, 0x0, 0x0)
	/workspace/controllers/globaldnsrecord/endpointstatus.go:41 +0x174
github.com/redhat-cop/global-load-balancer-operator/controllers/globaldnsrecord.(*GlobalDNSRecordReconciler).getAWSTrafficPolicyDocument(0xc0001ae230, 0xc000634340, 0xc006eecfd0, 0xc002b18338, 0x0, 0x0, 0x0, 0x0)
	/workspace/controllers/globaldnsrecord/route53provider.go:520 +0x29aa
github.com/redhat-cop/global-load-balancer-operator/controllers/globaldnsrecord.(*GlobalDNSRecordReconciler).createAWSTrafficPolicy(0xc0001ae230, 0xc000634340, 0xc006eecfd0, 0xc002b18338, 0x0, 0x0, 0x1a7cac0, 0x17fe51f)
	/workspace/controllers/globaldnsrecord/route53provider.go:322 +0x85
github.com/redhat-cop/global-load-balancer-operator/controllers/globaldnsrecord.(*GlobalDNSRecordReconciler).ensureRoute53TrafficPolicy(0xc0001ae230, 0xc000634340, 0xc002b18338, 0xc006eecfd0, 0xc002b18338, 0x0, 0x0, 0xc00430f7c0)
	/workspace/controllers/globaldnsrecord/route53provider.go:201 +0xa45
github.com/redhat-cop/global-load-balancer-operator/controllers/globaldnsrecord.(*GlobalDNSRecordReconciler).createRoute53Record(0xc0001ae230, 0x1ecd620, 0xc000930270, 0xc000634340, 0xc006eca160, 0xc006eecfd0, 0xa, 0xc006ef62c0, 0x1d, 0xc006ec0db0)
	/workspace/controllers/globaldnsrecord/route53provider.go:45 +0x371
github.com/redhat-cop/global-load-balancer-operator/controllers/globaldnsrecord.(*GlobalDNSRecordReconciler).Reconcile(0xc0001ae230, 0x1ecd620, 0xc000930270, 0xc00034b160, 0x1d, 0xc00034b0e0, 0x20, 0xc000930270, 0x40a1bf, 0xc000030000, ...)
	/workspace/controllers/globaldnsrecord/globaldnsrecord_controller.go:166 +0xf17
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler(0xc000ba2140, 0x1ecd560, 0xc00035a000, 0x1a41f00, 0xc000ce05e0)
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.7.0/pkg/internal/controller/controller.go:263 +0x317
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0xc000ba2140, 0x1ecd560, 0xc00035a000, 0x0)
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.7.0/pkg/internal/controller/controller.go:235 +0x205
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1.1(0x1ecd560, 0xc00035a000)
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.7.0/pkg/internal/controller/controller.go:198 +0x4a
k8s.io/apimachinery/pkg/util/wait.JitterUntilWithContext.func1()
	/go/pkg/mod/k8s.io/apimachinery@v0.20.0/pkg/util/wait/wait.go:185 +0x37
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0xc0009db750)
	/go/pkg/mod/k8s.io/apimachinery@v0.20.0/pkg/util/wait/wait.go:155 +0x5f
k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0xc006eedf50, 0x1e8f720, 0xc0009300c0, 0xc00035a001, 0xc000844300)
	/go/pkg/mod/k8s.io/apimachinery@v0.20.0/pkg/util/wait/wait.go:156 +0xad
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc0009db750, 0x3b9aca00, 0x0, 0x1, 0xc000844300)
	/go/pkg/mod/k8s.io/apimachinery@v0.20.0/pkg/util/wait/wait.go:133 +0x98
k8s.io/apimachinery/pkg/util/wait.JitterUntilWithContext(0x1ecd560, 0xc00035a000, 0xc0009200a0, 0x3b9aca00, 0x0, 0x1cd4801)
	/go/pkg/mod/k8s.io/apimachinery@v0.20.0/pkg/util/wait/wait.go:185 +0xa6
k8s.io/apimachinery/pkg/util/wait.UntilWithContext(0x1ecd560, 0xc00035a000, 0xc0009200a0, 0x3b9aca00)
	/go/pkg/mod/k8s.io/apimachinery@v0.20.0/pkg/util/wait/wait.go:99 +0x57
created by sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.7.0/pkg/internal/controller/controller.go:195 +0x4e7
```

Kindly let me know, if you need any information.

Regards,
Hari

--- a/controllers/globaldnsrecord/endpointstatus.go
+++ b/controllers/globaldnsrecord/endpointstatus.go
@@ -111,6 +111,7 @@ func (r *GlobalDNSRecordReconciler) getEndPointStatus(context context.Context, e
 		client:         client,
 		service:        *service,
 		infrastructure: *infrastructure,
+		log:            r.Log.WithName(endpoint.GetKey()),
 	}
 
 	return &endpointStatus, nil

--- a/controllers/globaldnsrecord/service_reconciler.go
+++ b/controllers/globaldnsrecord/service_reconciler.go
@@ -35,7 +35,7 @@ func newServiceReconciler(mgr *remotemanager.RemoteManager, statusChange chan<- 
 	controllerName := endpoint.GetKey()
 
 	serviceReconciler := &ServiceReconciler{
-		ReconcilerBase: util.NewReconcilerBase(mgr.GetClient(), mgr.GetScheme(), mgr.GetConfig(), mgr.GetEventRecorderFor(controllerName)),
+		ReconcilerBase: util.NewReconcilerBase(mgr.GetClient(), mgr.GetScheme(), mgr.GetConfig(), mgr.GetEventRecorderFor(controllerName), mgr.GetAPIReader()),
 		statusChange:   statusChange,
 		remoteManager:  mgr,
 		parentClient:   parentClient,

--- a/controllers/globalroutediscovery/routereconciler.go
+++ b/controllers/globalroutediscovery/routereconciler.go
@@ -43,7 +43,7 @@ func (r *GlobalRouteDiscoveryReconciler) newRouteReconciler(mgr *remotemanager.R
 	controllerName := cluster.ClusterName
 
 	routeReconciler := &RouteReconciler{
-		ReconcilerBase: util.NewReconcilerBase(mgr.GetClient(), mgr.GetScheme(), mgr.GetConfig(), mgr.GetEventRecorderFor(controllerName)),
+		ReconcilerBase: util.NewReconcilerBase(mgr.GetClient(), mgr.GetScheme(), mgr.GetConfig(), mgr.GetEventRecorderFor(controllerName), mgr.GetAPIReader()),
 		statusChange:   reconcileEventChannel,
 		remoteManager:  mgr,
 		parentClient:   parentClient,

--- a/go.mod
+++ b/go.mod
@@ -6,12 +6,12 @@ require (
 	github.com/aws/aws-sdk-go v1.36.15
 	github.com/fatih/set v0.2.1
 	github.com/go-logr/logr v0.3.0
-	github.com/gojp/goreportcard v0.0.0-20201106142952-232d912e513e // indirect
+	github.com/kr/pretty v0.2.1 // indirect
 	github.com/onsi/ginkgo v1.14.1
 	github.com/onsi/gomega v1.10.2
 	github.com/openshift/api v0.0.0-20201103184615-27004eede929
 	github.com/openshift/cloud-credential-operator v0.0.0-20201217140048-e0ae005686d1
-	github.com/redhat-cop/operator-utils v1.0.1
+	github.com/redhat-cop/operator-utils v1.1.2
 	github.com/scylladb/go-set v1.0.2
 	k8s.io/api v0.20.0
 	k8s.io/apimachinery v0.20.0

--- a/main.go
+++ b/main.go
@@ -100,21 +100,21 @@ func main() {
 	}
 
 	if err = (&globaldnsrecord.GlobalDNSRecordReconciler{
-		ReconcilerBase: util.NewReconcilerBase(mgr.GetClient(), mgr.GetScheme(), mgr.GetConfig(), mgr.GetEventRecorderFor("GlobalDNSRecord_controller")),
+		ReconcilerBase: util.NewReconcilerBase(mgr.GetClient(), mgr.GetScheme(), mgr.GetConfig(), mgr.GetEventRecorderFor("GlobalDNSRecord_controller"), mgr.GetAPIReader()),
 		Log:            ctrl.Log.WithName("controllers").WithName("GlobalDNSRecord"),
 	}).SetupWithManager(mgr); err != nil {
 		setupLog.Error(err, "unable to create controller", "controller", "GlobalDNSRecord")
 		os.Exit(1)
 	}
 	if err = (&globalroutediscovery.GlobalRouteDiscoveryReconciler{
-		ReconcilerBase: util.NewReconcilerBase(mgr.GetClient(), mgr.GetScheme(), mgr.GetConfig(), mgr.GetEventRecorderFor("GlobalRouteDiscovery_controller")),
+		ReconcilerBase: util.NewReconcilerBase(mgr.GetClient(), mgr.GetScheme(), mgr.GetConfig(), mgr.GetEventRecorderFor("GlobalRouteDiscovery_controller"), mgr.GetAPIReader()),
 		Log:            ctrl.Log.WithName("controllers").WithName("GlobalRouteDiscovery"),
 	}).SetupWithManager(mgr); err != nil {
 		setupLog.Error(err, "unable to create controller", "controller", "GlobalRouteDiscovery")
 		os.Exit(1)
 	}
 	if err = (&globaldnszone.GlobalDNSZoneReconciler{
-		ReconcilerBase: util.NewReconcilerBase(mgr.GetClient(), mgr.GetScheme(), mgr.GetConfig(), mgr.GetEventRecorderFor("GlobalDNSZone_controller")),
+		ReconcilerBase: util.NewReconcilerBase(mgr.GetClient(), mgr.GetScheme(), mgr.GetConfig(), mgr.GetEventRecorderFor("GlobalDNSZone_controller"), mgr.GetAPIReader()),
 		Log:            ctrl.Log.WithName("controllers").WithName("GlobalDNSZone"),
 	}).SetupWithManager(mgr); err != nil {
 		setupLog.Error(err, "unable to create controller", "controller", "GlobalDNSZone")

